lib/hl7/datetime_components.rb

```diff
@@ -0,0 +1,125 @@
+require 'active_support/core_ext/time'
+
+module HL7
+  class DatetimeComponents
+    attr_reader :year, :month, :day, :hour, :minute, :second, :fraction, :offset_seconds
+
+    def initialize(year, month = nil, day = nil, hour = nil, minute = nil, second = nil, fraction = nil, offset_seconds: nil)
+      @year = year
+      @month = month
+      @day = day
+      @hour = hour
+      @minute = minute
+      @second = second
+      @fraction = fraction
+      @offset_seconds = offset_seconds
+      validate!
+    end
+
+    def to_time(zone:)
+      full_seconds = second.to_i + fraction.to_f
+
+      if offset_seconds.nil?
+        Time.use_zone(zone) do
+          Time.zone.local(year, month, day, hour, minute, full_seconds)
+        end
+      else
+        Time.new(year, month, day, hour, minute, full_seconds, offset_seconds).
+          in_time_zone(zone)
+      end
+    end
+
+    def ==(other)
+      other.is_a?(DatetimeComponents) &&
+        self.to_time(zone: "UTC").eql?(other.to_time(zone: "UTC"))
+    end
+
+    private
+
+    def validate!
+      # year is not allowed to be nil
+      AttrSpec.new(:year, -9999..9999).validate_non_nil!(year)
+
+      # From month down to fraction-of-second, we validate the fields until we come
+      # to a nil value. Once we've found a nil, we ensure that all of the rest of
+      # the fields are also nil.
+      [
+        AttrSpec.new(:month, 1..12),
+        AttrSpec.new(:day, -> { 1..Time.days_in_month(month, year) }),
+        AttrSpec.new(:hour, 0..23),
+        AttrSpec.new(:minute, 0..59),
+        AttrSpec.new(:second, 0..59),
+        AttrSpec.new(:fraction, 0...1, :real?)
+      ].reduce(nil) do |first_nil_attr, spec|
+        value = public_send(spec.attr)
+        spec.validate!(value, first_nil_attr)
+      end
+
+      # offset_seconds is allowed to be nil or not
+      AttrSpec.new(:offset_seconds, -64800..64800).validate!(offset_seconds, nil)
+    end
+
+    class AttrSpec
+      attr_reader :attr
+
+      # The range_or_proc is either a Range or a callable that takes
+      # zero arguments and returns a Range. This allows Ranges to
+      # be lazily instantiated where necessary.
+      def initialize(attr, range_or_proc, predicate = :integer?)
+        @attr = attr
+        @range_or_proc = range_or_proc
+        @predicate = predicate
+      end
+
+      # This method always returns the first nil attribute that
```

Comment:
Maybe this would be more clear if I wrote: "... always returns the name of the first non-nil attribute we've encountred..."

And the parameter could be called `first_nil_attr_name`. Just to reduce the possibility of a confusion with the attribute _value_.